### PR TITLE
feat(RHTAPREL-435): refactor slack notification

### PR DESCRIPTION
**Pull request by scoheb**
- revert to using webhook integration.
- using the postMessage API in slack requires a bot scope that sometimes can be hard to obtain by IT security teams.
- secret name and key name within secret are now specified within the data structure.

2nd attempt - using webhook

This is re-creation of #258 since the source repo/branch were deleted my mistake.

### CODE REVIEW

File: pipelines/rhtap-service-push/rhtap-service-push.yaml [hunk 1]
```diff
@@ -279,24 +301,31 @@ spec:
...
         - push-sbom-to-pyxis
   finally:
-    - name: slack-webhook-notification
+    - name: send-slack-notification
       taskRef:
         resolver: "git"
         params:
           - name: url
-            value: https://github.com/redhat-appstudio/release-service-catalog.git
+            value: https://github.com/scoheb/release-service-catalog.git
```
Comment (davidmogar) [inline pipelines/rhtap-service-push/rhtap-service-push.yaml:309]: Why pointing to your fork when you are adding here the task?

Comment (scoheb) [reply pipelines/rhtap-service-push/rhtap-service-push.yaml:309]: so I can test the PR from a RPA before merging.

I will change the location before merge.

File: tasks/collect-slack-notification-params/README.md [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+Tekton task that collects slack configuration options from Release CRs and the data file. A formatted message is emitted
+as a task result for downstream tasks to use.
+
+## Parameters
+
+| Name                 | Description                                                                                                  | Optional | Default value                                                      |
+|----------------------|--------------------------------------------------------------------------------------------------------------|----------|--------------------------------------------------------------------|
+| release              | The namespaced name (namespace/name) of the Release custom resource initiating this pipeline execution       | No       | -                                                                  |
+| releaseplan          | The namespaced name (namespace/name) of the releasePlan                                                      | No       | -                                                                  |
```
Comment (mmalina) [inline tasks/collect-slack-notification-params/README.md:11]: It seems that RP and RPA are not actually there? Neither in the task yaml, nor in the call of the task in the pipeline.

File: tasks/collect-slack-notification-params/collect-slack-notification-params.yaml [hunk 3]
```diff
@@ -0,0 +1,185 @@
+---
+apiVersion: tekton.dev/v1
+kind: Task
+metadata:
+  name: collect-slack-notification-params
+  labels:
+    app.kubernetes.io/version: "0.0.1"
```
Comment (mmalina) [inline tasks/collect-slack-notification-params/collect-slack-notification-params.yaml:7]: In the past we would start with 0.1.0, but I think based on a recent comment by @davidmogar we should start with 1.0.0 (as anything in production should be at least 1.0.0).

File: tasks/collect-slack-notification-params/collect-slack-notification-params.yaml [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        fi
+
+        if [ $(jq '.slack | has("slack-notification-secret")' $DATA_FILE) == false ] ; then
+            echo "No secret name provided via 'slack.slack-notification-secret' key in Data."
+            echo -n "" > $(results.message.path)
+            echo -n "" > $(results.slack-notification-secret.path)
+            echo -n "" > $(results.slack-notification-secret-keyname.path)
+            exit
+        fi
+        slack_notification_secret=$(jq -r '.slack."slack-notification-secret"' $DATA_FILE)
```
Comment (mmalina) [inline tasks/collect-slack-notification-params/collect-slack-notification-params.yaml:74]: I was gonna complain about the `""` - it seems weird. But wow, it's actually needed. I learnt something new today :)

Apparently without it, jq will treat `-` as a subtraction.

File: tasks/collect-slack-notification-params/collect-slack-notification-params.yaml [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+            echo -n "" > $(results.slack-notification-secret-keyname.path)
+            exit
+        fi
+        slack_notification_secret_keyname=$(jq -r '.slack."slack-webhook-notification-secret-keyname"' $DATA_FILE)
+        echo -n "${slack_notification_secret_keyname}" > $(results.slack-notification-secret-keyname.path)
+
+        origin_namespace=$(jq -jr '.metadata.namespace' $RELEASE_FILE)
+        target_namespace=$(jq -r '.status.target' $RELEASE_FILE)
+
+        origin_workspace=$(echo $origin_namespace | sed 's/-tenant//g')
```
Comment (mmalina) [inline tasks/collect-slack-notification-params/collect-slack-notification-params.yaml:89]: I would probably suggest using `${origin_workspace%-tenant}`, see "Substring Removal" here: https://tldp.org/LDP/abs/html/string-manipulation.html
But if you prefer to use `sed`, then at least add `$` to limit this substitution to the end of the string: `sed 's/-tenant$//g'`

The same applies to the next line as well.

File: tasks/collect-slack-notification-params/collect-slack-notification-params.yaml [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+        if [ $(jq '.slack | has("slack-webhook-notification-secret-keyname")' $DATA_FILE) == false ] ; then
+            echo "No secret key name provided via 'slack.slack-webhook-notification-secret-keyname' key in Data."
+            echo -n "" > $(results.message.path)
+            echo -n "" > $(results.slack-notification-secret-keyname.path)
+            exit
+        fi
+        slack_notification_secret_keyname=$(jq -r '.slack."slack-webhook-notification-secret-keyname"' $DATA_FILE)
+        echo -n "${slack_notification_secret_keyname}" > $(results.slack-notification-secret-keyname.path)
+
+        origin_namespace=$(jq -jr '.metadata.namespace' $RELEASE_FILE)
```
Comment (mmalina) [inline tasks/collect-slack-notification-params/collect-slack-notification-params.yaml:86]: I think the `-j` is not needed. I actually didn't know it existed until now. But [the documentation](https://jqlang.github.io/jq/manual/) says: `Like -r but jq won't print a newline after each output.`

So it doesn't really make sense to use both. And AFAIK, the newline gets stripped in this assignment anyway, so `-r` is enough in most cases.

This applies to many other places below as well.

File: tasks/send-slack-notification/send-slack-notification.yaml [hunk 7]
```diff
@@ -0,0 +1,90 @@
+---
+apiVersion: tekton.dev/v1
+kind: Task
+metadata:
+  labels:
+    app.kubernetes.io/version: "0.0.1"
```
Comment (mmalina) [inline tasks/send-slack-notification/send-slack-notification.yaml:6]: Same thing with version as the other task - should probably be 1.0.0.

File: tasks/send-slack-notification/README.md
Comment (mmalina) [inline tasks/send-slack-notification/README.md:1]: Shouldn't we remove the existing task at the same time? Or is it/will it be used somewhere else?

File: tasks/send-slack-notification/send-slack-notification.yaml [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    Sends message to Slack using postMessage API
+  params:
+    - name: message
+      description: Message to be sent
+    - name: tasksStatus
+      description: status of tasks execution
+    - name: secretName
+      description: |
+        Name of secret which contains authentication token for app
+    - name: secretKeyName
```
Comment (mmalina) [inline tasks/send-slack-notification/send-slack-notification.yaml:22]: This one is missing from the README.

File: tasks/send-slack-notification/send-slack-notification.yaml [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+  steps:
+    - name: send-message
+      image: >-
+        quay.io/hacbs-release/release-utils@sha256:5733ece907aa70d6ebced36484c936731e8b27bfcf87fed226a0ecf95047a6b8
+      volumeMounts:
+        - name: slack-token
+          mountPath: "/etc/secrets"
+          readOnly: true
+      env:
+        - name: HOME
```
Comment (mmalina) [inline tasks/send-slack-notification/send-slack-notification.yaml:38]: What's this for? Seems to be unused?

File: tasks/send-slack-notification/README.md
Comment (scoheb) [reply tasks/send-slack-notification/README.md:1]: I will once this is merged.

File: tasks/collect-slack-notification-params/README.md (hunk 2 above)
Comment (scoheb) [reply tasks/collect-slack-notification-params/README.md]: fixed

File: tasks/collect-slack-notification-params/collect-slack-notification-params.yaml (hunk 5 above)
Comment (scoheb) [reply tasks/collect-slack-notification-params/collect-slack-notification-params.yaml]: Fixed using the $ with sed

File: tasks/collect-slack-notification-params/collect-slack-notification-params.yaml (hunk 6 above)
Comment (scoheb) [reply tasks/collect-slack-notification-params/collect-slack-notification-params.yaml]: I removed the '-j'. thanks

File: tasks/send-slack-notification/send-slack-notification.yaml (hunk 7 above)
Comment (scoheb) [reply tasks/send-slack-notification/send-slack-notification.yaml]: fixed

File: tasks/send-slack-notification/send-slack-notification.yaml (hunk 9 above)
Comment (scoheb) [reply tasks/send-slack-notification/send-slack-notification.yaml]: removed

File: tasks/send-slack-notification/README.md
Comment (mmalina) [reply tasks/send-slack-notification/README.md:1]: ok

File: tasks/send-slack-notification/README.md [hunk 10]
```diff
@@ -0,0 +1,11 @@
+# send-slack-notification task
+
+Sends message to Slack using postMessage API
+
+## Parameters
+| Name            | Description                                                | Optional | Default Value             |
+|-----------------|------------------------------------------------------------|----------|---------------------------|
+| message         | Message to be sent                                         | false    |                           |
```
Comment (mmalina) [inline tasks/send-slack-notification/README.md:8]: The convention in this repo is to say "Yes" or "No" in the "Optional" column. So we should probably stick with that.

File: tasks/send-slack-notification/send-slack-notification.yaml (hunk 8 above)
Comment (mmalina) [reply tasks/send-slack-notification/send-slack-notification.yaml:22]: I can see this was added.

File: tasks/send-slack-notification/README.md (hunk 10 above)
Comment (scoheb) [reply tasks/send-slack-notification/README.md]: Fixed

**Review by mmalina — COMMENTED** [review]
LGTM, except for the scoheb fork references.

**Comment by scoheb** [comment]
fixed `scoheb` fork reference

**Review by mmalina — APPROVED** [review]

**Pull request merged**